crypto/chacha/asm/chacha-armv8-sve.pl

```diff
@@ -0,0 +1,840 @@
+___
+	if (@_) {
+		&SVE2_XAR($bits,@_);
+	}
+}
+
+sub SVE_QR_GROUP() {
+	my $have_sve2 = shift;
+	my ($a0,$b0,$c0,$d0,$a1,$b1,$c1,$d1,$a2,$b2,$c2,$d2,$a3,$b3,$c3,$d3) = @_;
+
+	&SVE_ADD($a0,$b0,$a1,$b1,$a2,$b2,$a3,$b3);
+	&SVE_EOR($d0,$a0,$d1,$a1,$d2,$a2,$d3,$a3);
+	&SVE_REV16($d0,$d1,$d2,$d3);
+
+	&SVE_ADD($c0,$d0,$c1,$d1,$c2,$d2,$c3,$d3);
+	if ($have_sve2 == 0) {
+		&SVE_EOR($b0,$c0,$b1,$c1,$b2,$c2,$b3,$c3);
+		&SVE_LSL(12,0,$b0,$b1,$b2,$b3);
+		&SVE_LSR(20,$b0,$b1,$b2,$b3);
+		&SVE_ORR(0,$b0,$b1,$b2,$b3,);
+	} else {
+		&SVE2_XAR(12,$b0,$c0,$b1,$c1,$b2,$c2,$b3,$c3);
+	}
+
+	&SVE_ADD($a0,$b0,$a1,$b1,$a2,$b2,$a3,$b3);
+	&SVE_EOR($d0,$a0,$d1,$a1,$d2,$a2,$d3,$a3);
+	&SVE_ROT8($d0,$d1,$d2,$d3);
+
+	&SVE_ADD($c0,$d0,$c1,$d1,$c2,$d2,$c3,$d3);
+	if ($have_sve2 == 0) {
+		&SVE_EOR($b0,$c0,$b1,$c1,$b2,$c2,$b3,$c3);
+		&SVE_LSL(7,0,$b0,$b1,$b2,$b3);
+		&SVE_LSR(25,$b0,$b1,$b2,$b3);
+		&SVE_ORR(0,$b0,$b1,$b2,$b3);
+	} else {
+		&SVE2_XAR(7,$b0,$c0,$b1,$c1,$b2,$c2,$b3,$c3);
+	}
+}
+
+sub SVE_INNER_BLOCK() {
+$code.=<<___;
+	//cbnz $sve2flag, 10f
+___
+	&SVE_QR_GROUP(0,0,4,8,12,1,5,9,13,2,6,10,14,3,7,11,15);
+	&SVE_QR_GROUP(0,0,5,10,15,1,6,11,12,2,7,8,13,3,4,9,14);
+$code.=<<___;
+	// SVE 2 not enabled until hardware available
+#if 0
+	b 11f
+10:
+___
+#	&SVE_QR_GROUP(1,0,4,8,12,1,5,9,13,2,6,10,14,3,7,11,15);
+#	&SVE_QR_GROUP(1,0,5,10,15,1,6,11,12,2,7,8,13,3,4,9,14);
+$code.=<<___;
+11:
+#endif
+___
+}
+
+{{{
+my ($dlen,$rsize,$tmp) = ("x10","x11","x12");
+
+sub load() {
+	my $x0 = shift;
+	my $x1 = shift;
+	my $x2 = shift;
+	my $x3 = shift;
+	my $x4 = shift;
+	my $x5 = shift;
+	my $x6 = shift;
+	my $x7 = shift;
+
+$code.=<<___;
+	ld1w	{$x0.s},p0/z,[$inp]
+	ld1w	{$x1.s},p0/z,[$inp, #1, MUL VL]
+	ld1w	{$x2.s},p0/z,[$inp, #2, MUL VL]
+	ld1w	{$x3.s},p0/z,[$inp, #3, MUL VL]
+	ld1w	{$x4.s},p0/z,[$inp, #4, MUL VL]
+	ld1w	{$x5.s},p0/z,[$inp, #5, MUL VL]
+	ld1w	{$x6.s},p0/z,[$inp, #6, MUL VL]
+	ld1w	{$x7.s},p0/z,[$inp, #7, MUL VL]
+	addvl	$inp,$inp,#8
+___
+}
+
+sub store() {
+	my $x0 = shift;
+	my $x1 = shift;
+	my $x2 = shift;
+	my $x3 = shift;
+	my $x4 = shift;
+	my $x5 = shift;
+	my $x6 = shift;
+	my $x7 = shift;
+
+$code.=<<___;
+	st1w	{$x0.s},p0,[$outp]
+	st1w	{$x1.s},p0,[$outp, #1, MUL VL]
+	st1w	{$x2.s},p0,[$outp, #2, MUL VL]
+	st1w	{$x3.s},p0,[$outp, #3, MUL VL]
+	st1w	{$x4.s},p0,[$outp, #4, MUL VL]
+	st1w	{$x5.s},p0,[$outp, #5, MUL VL]
+	st1w	{$x6.s},p0,[$outp, #6, MUL VL]
+	st1w	{$x7.s},p0,[$outp, #7, MUL VL]
+	addvl	$outp,$outp,#8
+___
+}
+
+sub transpose() {
+	my $xa = shift;
+	my $xb = shift;
+	my $xc = shift;
+	my $xd = shift;
+
+$code.=<<___;
+	zip1	$xt8.s,$xa.s,$xb.s
+	zip2	$xt9.s,$xa.s,$xb.s
+	zip1	$xt10.s,$xc.s,$xd.s
+	zip2	$xt11.s,$xc.s,$xd.s
+	zip1	$xa.d,$xt8.d,$xt10.d
+	zip2	$xb.d,$xt8.d,$xt10.d
+	zip1	$xc.d,$xt9.d,$xt11.d
+	zip2	$xd.d,$xt9.d,$xt11.d
+___
+}
+
+sub add_states() {
+	my ($tmpw0,$tmpw1,$tmpw2,$tmpw3) = ("w10","w11","w12","w13");
+
+$code.=<<___;
+	ldp	$tmpw0,$tmpw1,[$state]
+	ldp	$tmpw2,$tmpw3,[$state,#8]
```

Comment:
Small optimization opportunity:
I notice currently z13-z15 are not being used and SVE codepath is only used when vector length >= 256b.

You could load a copy of state (64 bytes fits into 2x256b vectors) into z13, z14 once at the start of SVE code, and in these places you want to dup the state (add_states and SVE_LOAD_STATES), use [DUP (indexed)](https://developer.arm.com/documentation/ddi0602/2022-03/SVE-Instructions/DUP--indexed---Broadcast-indexed-element-to-vector--unpredicated--?lang=en), rather than loading into scalars and duplicating from scalar to vector registers.

Similarly, you could only use the INDEX instruction once at the beginning, writing to z15, and update it with [INCW (vector)](https://developer.arm.com/documentation/ddi0602/2022-03/SVE-Instructions/INCD--INCH--INCW--vector---Increment-vector-by-multiple-of-predicate-constraint-element-count-?lang=en), rather than moving the counter between scalar and vector registers.

Reply:
changes to INDEX makes code slightly faster, also the code looks cleaner, so I pushed the new changes.

Reply:
I'm guessing you also already tried [LD1RW](https://developer.arm.com/documentation/ddi0602/2022-03/SVE-Instructions/LD1RW--Load-and-broadcast-unsigned-word-to-vector-?lang=en) in place of LDP (scalar) + DUP (scalar), or DUP (indexed)?

Reply:
ld1rw will slow down the code. it seems the mix of scalar load and DUP are the best combination so far.